src/ray/gcs/gcs_server/gcs_resource_report_poller.cc

```diff
@@ -0,0 +1,148 @@
+    // thread in rapid succession.
+    polling_service_.stop();
+    if (polling_thread_->joinable()) {
+      polling_thread_->join();
+    }
+  }
+}
+
+void GcsResourceReportPoller::HandleNodeAdded(
+    const std::shared_ptr<rpc::GcsNodeInfo> &node_info) {
+  absl::MutexLock guard(&mutex_);
+
+  auto state = std::make_shared<PullState>();
+  state->node_id = NodeID::FromBinary(node_info->node_id());
+
+  const NodeID &node_id = state->node_id;
+
+  RAY_CHECK(!nodes_.count(node_id)) << "Node with id: " << node_id << " was added twice!";
+
+  state->address.set_raylet_id(node_info->node_id());
+  state->address.set_ip_address(node_info->node_manager_address());
+  state->address.set_port(node_info->node_manager_port());
+
+  state->last_pull_time = -1;
+  state->next_pull_time = get_current_time_milli_();
+
+  nodes_[node_id] = state;
+  to_pull_queue_.push_front(state);
+  RAY_LOG(DEBUG) << "Node was added with id: " << node_id;
+
+  polling_service_.post([this]() { TryPullResourceReport(); });
+}
+
+void GcsResourceReportPoller::HandleNodeRemoved(
+    const std::shared_ptr<rpc::GcsNodeInfo> &node_info) {
+  NodeID node_id = NodeID::FromBinary(node_info->node_id());
+  {
+    absl::MutexLock guard(&mutex_);
+    RAY_CHECK(false);
+    nodes_.erase(node_id);
+  }
+}
+
+void GcsResourceReportPoller::Tick() { TryPullResourceReport(); }
+
+void GcsResourceReportPoller::TryPullResourceReport() {
+  absl::MutexLock guard(&mutex_);
+  int64_t cur_time = get_current_time_milli_();
+
+  RAY_LOG(DEBUG) << "Trying to pull inflight_pulls " << inflight_pulls_ << "/"
+                 << max_concurrent_pulls_ << ", queue size: " << to_pull_queue_.size();
+  while (inflight_pulls_ < max_concurrent_pulls_ && !to_pull_queue_.empty()) {
+    auto to_pull = to_pull_queue_.front();
+    if (cur_time < to_pull->next_pull_time) {
+      break;
+    }
+
+    to_pull_queue_.pop_front();
+
+    if (!nodes_.count(to_pull->node_id)) {
+      RAY_LOG(DEBUG)
+          << "Update finished, but node was already removed from the cluster. Ignoring.";
+      continue;
+    }
+
+    PullResourceReport(to_pull);
+  }
+}
+
+void GcsResourceReportPoller::PullResourceReport(const std::shared_ptr<PullState> state) {
+  inflight_pulls_++;
+
+  request_report_(
+      state->address, raylet_client_pool_,
+      [this, state](const Status &status, const rpc::RequestResourceReportReply &reply) {
+        if (status.ok()) {
+          // TODO (Alex): This callback is always posted onto the main thread. Since most
+          // of the work is in the callback we should move this callback's execution to
+          // the polling thread. We will need to implement locking once we switch threads.
+          handle_resource_report_(reply.resources());
+          polling_service_.post([this, state]() { NodeResourceReportReceived(state); });
+        } else {
+          RAY_LOG(INFO) << "Couldn't get resource request from raylet " << state->node_id
```

Comment:
Don't we need to update inflight_pulls_ and process the queue again here?

Can you add this case to the unit tests?

Reply:
Also, I think we still want to schedule the next pull, right? We don't want to stop pulling unless the node is actually removed.

Reply:
I think both of those things are done in `NodeResourceReportReceived`, in the spirit of minimizing the logic in the callback. 

Btw I think `TestBasic` Does cover this, since it continues to pull after the first pulls complete, without a timer tick every happening.

Reply:
But doesn't `NodeResourceReportReceived` only get called in the status==OK case?

Reply:
Oh I see, you're saying we should do the continuation logic regardless of status right? (will add a test for this, just want to confirm the semantics first)

Reply:
Yes.